insonmnia/dwh/server_test.go

```diff
@@ -18,47 +26,193 @@ import (
+	if err != nil {
+		cli.Close()
+		return nil, "", fmt.Errorf("failed to pull postgres image: %s", err)
+	}
+	io.Copy(os.Stdout, reader)
+
+	containerCfg := &container.Config{
+		Image:        "postgres",
+		ExposedPorts: nat.PortSet{"5432": struct{}{}},
+	}
+	hostCfg := &container.HostConfig{
+		PortBindings: map[nat.Port][]nat.PortBinding{
+			nat.Port("5432"): {{HostIP: "localhost", HostPort: postgresPort}},
+		},
+	}
+	resp, err := cli.ContainerCreate(ctx, containerCfg, hostCfg, nil, "dwh-postgres-test")
+	if err != nil {
+		cli.Close()
+		return nil, "", fmt.Errorf("failed to create container: %s", err)
+	}
+
+	if err := cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{}); err != nil {
+		cli.Close()
+		return nil, "", fmt.Errorf("failed to start container: %s", err)
+	}
+
+	return cli, resp.ID, nil
+}
+
+func checkPostgresReadiness(containerID string) error {
+	var (
+		err        error
+		numRetries = 10
+	)
+	for ; numRetries > 0; numRetries-- {
+		cmd := exec.Command("docker", "exec", containerID, "pg_isready")
+		out, err := cmd.CombinedOutput()
+		if err == nil {
+			if strings.Contains(string(out), "accepting connections") {
+				return nil
+			}
+		}
+		fmt.Printf("postgres container not ready, %d retries left\n", numRetries)
+		time.Sleep(time.Second)
+	}
+
+	return fmt.Errorf("failed to connect to postgres container: %v", err)
+}
+
+func setupDB() error {
+	db, err := sql.Open("postgres", serviceConnString)
+	if err != nil {
+		return fmt.Errorf("failed to connect to template1: %s", err)
+	}
+	defer db.Close()
+
+	for _, dbName := range []string{globalDBName, monitorDBName} {
+		if _, err := db.Exec(fmt.Sprintf("DROP DATABASE IF EXISTS %s", dbName)); err != nil {
+			return fmt.Errorf("failed to preliminarily drop database: %s", err)
+		}
+	}
+
+	if _, err := db.Exec(fmt.Sprintf("DROP USER IF EXISTS %s", dbUser)); err != nil {
+		return fmt.Errorf("failed to preliminarily drop user: %s", err)
+	}
+
+	if _, err = db.Exec(fmt.Sprintf("CREATE USER %s WITH PASSWORD '%s'", dbUser, dbUserPassword)); err != nil {
+		return fmt.Errorf("failed to create user: %v", err)
+	}
+
+	for _, dbName := range []string{globalDBName, monitorDBName} {
+		if _, err := db.Exec(fmt.Sprintf("CREATE DATABASE %s OWNER dwh_tester", dbName)); err != nil {
+			return fmt.Errorf("failed to create database: %s", err)
+		}
+	}
+
+	return nil
+}
+
+func tearDownDB() error {
+	db, err := sql.Open("postgres", serviceConnString)
+	if err != nil {
+		return fmt.Errorf("failed to connect to template1: %s", err)
+	}
+	defer db.Close()
+
+	for _, dbName := range []string{globalDBName, monitorDBName} {
+		if _, err := db.Exec(fmt.Sprintf("DROP DATABASE IF EXISTS %s", dbName)); err != nil {
+			return fmt.Errorf("failed to drop database: %s", err)
+		}
+	}
+
+	return nil
+}
+
+func getConnString(database, user, password string) string {
+	return fmt.Sprintf("postgresql://localhost:15432/%s?user=%s&password=%s&sslmode=disable", database, user, password)
```

Comment:
maybe will be better to not hardcode port number and use a value from constants

Reply:
will do that in the next round of refactoring